### PR TITLE
:bug: Fix a bug in verify func of SMT to avoid panic for bitmap len

### LINKED ISSUE #106: Panic due to lack of proof's bitmap's length validation
### Expected behavior
It should have validation to avoid the panic. Also, add fuzz test to avoid unintended panic in the functions

### Actual behavior
In https://github.com/LiskHQ/lisk-db/blob/main/src/sparse_merkle_tree/smt.rs#LL1369C70-L1369C83, if the `query.pair.0` is shorter than the binary_bitmap length, it panics.

### Steps to reproduce
Add very long bitmap in the proof

### Which version(s) does this affect? (Environment, OS, etc...)
0.3.5-

### PR DESCRIPTION
**Pull request by hrmhatef**
### What was the problem?

This PR resolves #106 

### How was it solved?
- Add a simple validator in verify function of SMT

### How was it tested?

 - Add a simple unit test to check it.
 - All tests passed

### CODE REVIEW

File: src/sparse_merkle_tree/smt.rs [hunk 1]
```diff
@@ -1538,7 +1538,7 @@ impl SparseMerkleTree {
 
         // Check if all the query keys have the same length.
         for query in &proof.queries {
-            if query.pair.key().len() != key_length.into() {
+            if query.key().len() != key_length.into() || query.key().len() < query.bitmap.len() {
```
Comment (shuse2) [inline src/sparse_merkle_tree/smt.rs:1541]: i think here you need to compare with `binary_bitmap`. At this point `query.bitmap` is still in bytes format

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: src/sparse_merkle_tree/smt.rs (hunk 1 above)
Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs]: Done.

File: src/sparse_merkle_tree/smt.rs [hunk 2]
```diff
@@ -1362,24 +1364,31 @@ impl SparseMerkleTree {
...
-                binary_path,
-                QueryProofWithProof::new_with_pair(
-                    Arc::clone(&query.pair),
-                    &binary_bitmap,
-                    &[],
-                    &[],
-                ),
-            );
+        if !proof.queries.is_empty() {
+            let max_binary_bitmap_len = proof.queries[0].key().len() * 8; // converts key_length to bits
```
Comment (matjazv) [inline src/sparse_merkle_tree/smt.rs:1370]: Is this really maximum bitmap length?
Can some other query has longer key? (eg. `proof.queries[15]`)

Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs:1370]: The length of each key in queries are same also it is already check with `key_length` in `verify` function line `1550`

**Review by matjazv — APPROVED** [review]

File: src/sparse_merkle_tree/smt.rs [hunk 3]
```diff
@@ -1362,24 +1364,31 @@ impl SparseMerkleTree {
...
+            for query in &proof.queries {
+                let binary_bitmap = utils::strip_left_false(&utils::bytes_to_bools(&query.bitmap));
+                if binary_bitmap.len() > max_binary_bitmap_len {
+                    return Err(SMTError::InvalidBitmapLen);
+                }
+                let binary_path =
+                    utils::bytes_to_bools(query.key())[..binary_bitmap.len()].to_vec();
```
Comment (shuse2) [inline src/sparse_merkle_tree/smt.rs:1377]: wouldn't it be sufficient to just compare length of `utils::bytes_to_bools(query.key())` and `binary_bitmap.len()`?

Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs]: Thank you for that, it is better and more clear. Done.

File: src/sparse_merkle_tree/smt.rs [hunk 4]
```diff
@@ -2481,6 +2488,51 @@ mod tests {
...
+                    .iter()
+                    .map(|k| hex::decode(k).unwrap())
+                    .collect::<NestedVec>(),
+            )
+            .unwrap();
+
+        proof.queries[0].bitmap = Arc::new(vec![
+            1u8, 2, 3, 4, 1, 3, 3, 71, 3, 3, 71, 3, 3, 71, 3, 3, 71, 3, 3, 71, 3, 3, 71, 3, 3, 71,
+            3, 3, 71, 3, 3, 71, 3, 3, 7,
+        ]);
```
Comment (shuse2) [inline src/sparse_merkle_tree/smt.rs:2521]: i think we should check the cases
* 32 byte key with 32 byte bitmap
* 32 byte key with 33 byte bitmap

Comment (hrmhatef) [reply src/sparse_merkle_tree/smt.rs]: Done.

**Pull request merged**